src/stan/variational/advi.hpp

```diff
@@ -94,15 +89,30 @@ namespace stan {
        * @return   evidence lower bound (elbo)
        */
       double calc_ELBO(const Q& variational) const {
+        static const char* function =
+          "stan::variational::advi::calc_ELBO";
+
         double elbo = 0.0;
+        double energy_i;
         int dim = variational.dimension();
-
         Eigen::VectorXd zeta(dim);
+        int n_monte_carlo_drop = 0;
 
         for (int i = 0; i < n_monte_carlo_elbo_; ++i) {
           zeta = variational.sample(rng_);
-          // Accumulate log probability
-          elbo += model_.template log_prob<false, true>(zeta, print_stream_);
+          try {
+            energy_i = model_.template log_prob<false, true>(zeta,
+                                                             print_stream_);
+            stan::math::check_not_nan(function, "energy_i", energy_i);
+            stan::math::check_finite(function, "energy_i", energy_i);
+            elbo += energy_i;
+          } catch (std::exception& e) {
+            this->write_error_msg_(print_stream_, e);
+            i -= 1;
```

Comment:
instead of this, which takes a minute to process what you're doing, could you do something like:
```cpp
n = 0;
int n_monte_carlo_drop = 0;
while (n < n_monte_carlo_elbo_ && n_monte_carlo_drop < n_monte_carlo_elbo_) {
...
}
```

Or something like that. Decrementing the loop index within the loop is just tricky to figure out.